x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrationExecutor.java

```diff
@@ -0,0 +1,114 @@
+import org.elasticsearch.persistent.AllocatedPersistentTask;
+import org.elasticsearch.persistent.PersistentTaskState;
+import org.elasticsearch.persistent.PersistentTasksExecutor;
+import org.elasticsearch.tasks.TaskCancelledException;
+import org.elasticsearch.xpack.core.security.action.UpdateIndexMigrationVersionAction;
+import org.elasticsearch.xpack.core.security.support.SecurityMigrationTaskParams;
+
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.Executor;
+
+import static org.elasticsearch.xpack.security.support.SecurityMigrations.isEligibleSecurityMigration;
+
+public class SecurityMigrationExecutor extends PersistentTasksExecutor<SecurityMigrationTaskParams> {
+
+    private static final Logger logger = LogManager.getLogger(SecurityMigrationExecutor.class);
+    private final SecurityIndexManager securityIndexManager;
+    private final Client client;
+    private final FeatureService featureService;
+    private final ClusterService clusterService;
+    private final TreeMap<Integer, SecurityMigrations.SecurityMigration> migrationByVersion;
+
+    public SecurityMigrationExecutor(
+        String taskName,
+        Executor executor,
+        SecurityIndexManager securityIndexManager,
+        Client client,
+        FeatureService featureService,
+        ClusterService clusterService,
+        TreeMap<Integer, SecurityMigrations.SecurityMigration> migrationByVersion
+    ) {
+        super(taskName, executor);
+        this.securityIndexManager = securityIndexManager;
+        this.client = client;
+        this.featureService = featureService;
+        this.clusterService = clusterService;
+        this.migrationByVersion = migrationByVersion;
+    }
+
+    @Override
+    protected void nodeOperation(AllocatedPersistentTask task, SecurityMigrationTaskParams params, PersistentTaskState state) {
+        applyOutstandingMigrations(task, params.getMigrationVersion(), ActionListener.wrap((res) -> task.markAsCompleted(), (exception) -> {
+            logger.warn("Security migration failed: " + exception);
+            task.markAsFailed(exception);
+        }));
+    }
+
+    private void applyOutstandingMigrations(AllocatedPersistentTask task, int currentMigrationVersion, ActionListener<Void> listener) {
+        if (task.isCancelled()) {
+            listener.onFailure(new TaskCancelledException("Security migration task cancelled"));
+            return;
+        }
+        Map.Entry<Integer, SecurityMigrations.SecurityMigration> migrationEntry = migrationByVersion.higherEntry(currentMigrationVersion);
+
+        // Check if all nodes can support feature and that the cluster is on a compatible mapping version
+        if (migrationEntry != null
+            && isEligibleSecurityMigration(
+                securityIndexManager.indexMappingVersion(),
+                featureService,
+                clusterService.state(),
+                migrationEntry.getValue()
+            )) {
```

Comment:
Shouldn't we also check `clusterReadyForSecurityMigration` here (that the index is available for migration)? If the index is not available I think the task should be retried (which is what the code that enqueues the task in `Security.java` does).
More generally, if the reindexing fails (e.g. the indexing is not available for writing anymore), the persistent task fails, but it will be retried, no (i.e. it's OK to send a persistent task with the same ID as a failed task (I think that's the case, but asking to make sure))?

Reply:
Expanded on this and refactored it a lot. Thanks!